src/lockfile/wrapper.js

```diff
@@ -156,7 +160,8 @@ export default class Lockfile {
       invariant(ref, 'Package is missing a reference');
       invariant(remote, 'Package is missing a remote');
 
-      const seenPattern = remote.resolved && seen.get(remote.resolved);
+      const altRemoteKey = keyForRemote(remote);
```

Comment:
Yes this is better. It looks like there may be [one case](https://github.com/yarnpkg/yarn/blob/master/src/resolvers/exotics/hosted-git-resolver.js#L193-L197) where a remote may have resolved but a falsy hash, so I added a check for `remote.resolved` in `keyForRemote`.